modules/edu.gemini.seqexec.web/edu.gemini.seqexec.web.server/src/main/scala/edu/gemini/seqexec/web/server/security/FreeLDAPAuthenticationService.scala

```diff
@@ -0,0 +1,119 @@
+package edu.gemini.seqexec.web.server.security
+
+import com.unboundid.ldap.sdk._
+import edu.gemini.seqexec.web.common.UserDetails
+import edu.gemini.seqexec.web.server.security.AuthenticationService.AuthResult
+
+import scalaz.Scalaz._
+import scalaz._
+import scalaz.concurrent.Task
+import scalaz.effect.IO
+
+/**
+  * Definition of LDAP as a free monad algebra/interpreters
+  */
+object FreeLDAPAuthenticationService {
+  import UserDetails._
+  import LdapConnectionOps._
+
+  sealed trait LdapOp[A]
+  // Operations on ldap
+  object LdapOp {
+    // Operation to authenticate a user, It returns true if it works
+    case class AuthenticateOp(username: String, password: String) extends LdapOp[UID]
+    // Read the user display name
+    case class UserDisplayNameOp(uid: UID)                        extends LdapOp[DisplayName]
```

Comment:
`UID` is just an alias for `String`, so you can easily supply one. But it could be a proper type and then set the constructors in such a way that it could only be build inside `AuthenticateOp`

But as you said, it would be overkill in this case